### PR TITLE
Handle all valid URL path characters for rootURL/baseURL in tests-server

### LINKED ISSUE #5876: environment.baseURL is interpreted as regex by tests-server
Trying to use a `baseURL` of `/foo/++` doesn't work (you can't serve the app with ember `serve`).

According to [some specification](http://tools.ietf.org/html/rfc3986#section-3.3), `+` should be allowed in the URL segments.

# Steps to Reproduce
1. Create a new ember project: `ember new foo`
2. Change the baseURL in `config/environment.js` to `/foo/++`
3. Try to serve the app with `ember serve`

```
✗ ember serve
version: 2.4.2
Invalid regular expression: /^/foo/++/tests/: Nothing to repeat
SyntaxError: Invalid regular expression: /^/foo/++/tests/: Nothing to repeat
    at new RegExp (native)
    at TestsServerAddon.serverMiddleware (/Users/lsmith/projects/socialcode/ember-bug/foo/node_modules/ember-cli/lib/tasks/server/middleware/tests-server/index.js:20:21)
    at Class.<anonymous> (/Users/lsmith/projects/socialcode/ember-bug/foo/node_modules/ember-cli/lib/tasks/server/express-server.js:105:22)
    at /Users/lsmith/projects/socialcode/ember-bug/foo/node_modules/ember-cli/node_modules/promise-map-series/index.js:11:14
    at lib$rsvp$$internal$$tryCatch (/Users/lsmith/projects/socialcode/ember-bug/foo/node_modules/ember-cli/node_modules/rsvp/dist/rsvp.js:1036:16)
    at lib$rsvp$$internal$$invokeCallback (/Users/lsmith/projects/socialcode/ember-bug/foo/node_modules/ember-cli/node_modules/rsvp/dist/rsvp.js:1048:17)
    at /Users/lsmith/projects/socialcode/ember-bug/foo/node_modules/ember-cli/node_modules/rsvp/dist/rsvp.js:331:11
    at lib$rsvp$asap$$flush (/Users/lsmith/projects/socialcode/ember-bug/foo/node_modules/ember-cli/node_modules/rsvp/dist/rsvp.js:1198:9)
    at nextTickCallbackWith0Args (node.js:419:9)
    at process._tickCallback (node.js:348:13)
```

---

Output from ember version --verbose:
```
version: 2.4.2
http_parser: 2.5.0
node: 4.2.6
v8: 4.5.103.35
uv: 1.8.0
zlib: 1.2.8
ares: 1.10.1-DEV
icu: 56.1
modules: 46
openssl: 1.0.2e
os: darwin x64
```

# Excerpt from RFC 3986

```
path          = path-abempty    ; begins with "/" or is empty
                    / path-absolute   ; begins with "/" but not "//"
                    / path-noscheme   ; begins with a non-colon segment
                    / path-rootless   ; begins with a segment
                    / path-empty      ; zero characters

      path-abempty  = *( "/" segment )
      path-absolute = "/" [ segment-nz *( "/" segment ) ]
      path-noscheme = segment-nz-nc *( "/" segment )
      path-rootless = segment-nz *( "/" segment )
      path-empty    = 0<pchar>
```

```
segment       = *pchar
      segment-nz    = 1*pchar
      segment-nz-nc = 1*( unreserved / pct-encoded / sub-delims / "@" )
                    ; non-zero-length segment without any colon ":"

      pchar         = unreserved / pct-encoded / sub-delims / ":" / "@"
```

```
   pct-encoded   = "%" HEXDIG HEXDIG

   unreserved    = ALPHA / DIGIT / "-" / "." / "_" / "~"
   reserved      = gen-delims / sub-delims
   gen-delims    = ":" / "/" / "?" / "#" / "[" / "]" / "@"
   sub-delims    = "!" / "$" / "&" / "'" / "(" / ")"
                 / "*" / "+" / "," / ";" / "="
```

# Willing to Help

I am willing to help, even write code. But I am new to Ember! Any suggestions?

For now, I can just not use that character in my URL, but it would have been really cool, as in: `/c/++/`

### PR DESCRIPTION
**Pull request by les2**
The previous implementation used a regular expression for matching the
beginning of the request path against the baseURL + 'tests' but did
not escape the baseURL. Because RFC 3986 allows some characters that
are not valid in regexes, this caused problems when using them for
the baseURL.

Closes #5876

### CODE REVIEW

File: lib/tasks/server/middleware/tests-server/index.js [hunk 1]
```diff
@@ -41,7 +41,8 @@ TestsServerAddon.prototype.serverMiddleware = function(config) {
         var filePath = path.join(results.directory, assetPath);
 
         if (!existsSync(filePath) || !fs.statSync(filePath).isFile()) {
-          var newURL = baseURL + '/tests/index.html';
+          // N.B., `baseURL` will end with a slash as it went through `cleanBaseURL`
+          var newURL = baseURL + 'tests/index.html';
```
Comment (les2) [inline lib/tasks/server/middleware/tests-server/index.js]: This appeared to be an existing bug. Without this change I got a double `/`, making my assertions fail: `/braden/+//tests/index.html`.

File: tests/unit/tasks/server/middleware/tests-server-test.js [hunk 2]
```diff
@@ -20,6 +20,16 @@ describe('TestServerAddon', function () {
...
 
+    afterEach(function() {
+      nextWasCalled = false;
+      mockRequest = {
+        method: 'GET',
+        path: '',
+        url: 'http://example.com',
+        headers: {}
+      };
+    });
```
Comment (les2) [inline tests/unit/tasks/server/middleware/tests-server-test.js]: Assuming I needed to reset state between tests since `mockRequest`'ll get modified inside the middleware (it sets `.path` property specifically, but potentially anything could get dirty). If I was more ambitious, I'd just set everything up in a `beforeEach` block and then wouldn't need the `afterEach`. I think what I have here is the Minimum Viable Change. :)

File: tests/unit/tasks/server/middleware/tests-server-test.js [hunk 3]
```diff
@@ -47,5 +57,37 @@ describe('TestServerAddon', function () {
...
+          options: {
+            watcher: Promise.resolve({directory: 'nothing'}),
+            rootURL: '/grayson/+'
+          },
+          finally: function () {
+            expect(mockRequest.url).to.equal('/grayson/+/tests/index.html');
+            done();
+          }
+        });
+      });
```
Comment (les2) [inline tests/unit/tasks/server/middleware/tests-server-test.js]: I could also test some negative cases. I chose not to go crazy with testing (I've been known to do that). Let me know if you think we should add a few more cases (e.g., it should not modify the url if the file exists, if it doesn't match the test pattern, etc.).

**Comment by stefanpenner** [comment]
@rwjblue r?

**Comment by les2** [comment]
The AppVeyor problem is related to phantomjs-prebuilt not being installed (for the Environment: nodejs_version=4.2). Does not seem related to these changes.

retest this

```
> phantomjs-prebuilt@2.1.6 install C:\projects\ember-cli\node_modules\phantomjs-prebuilt
> node install.js
 
PhantomJS not found on PATH
Downloading https://bitbucket.org/ariya/phantomjs/downloads/phantomjs-2.1.1-windows.zip
Saving to C:\Users\appveyor\AppData\Local\Temp\1\phantomjs\phantomjs-2.1.1-windows.zip
Receiving...
 
Error requesting archive.
Status: 403
Request options: {
  "uri": "https://bitbucket.org/ariya/phantomjs/downloads/phantomjs-2.1.1-windows.zip",
  "encoding": null,
  "followRedirect": true,
  "headers": {
    "User-Agent": "npm/2.14.12 node/v4.2.6 win32 ia32"
  },
  "strictSSL": true
}
Response headers: {
  "x-amz-request-id": "19079527EFA0EDBF",
  "x-amz-id-2": "noVWU6FJkOI40JJ+Ynl4d3/xz/r9h8h9GgCwL6Q+d8/+7oW0FUb2xJig0ODzQ6xyxBc4wdpiZws=",
  "content-type": "application/xml",
  "transfer-encoding": "chunked",
  "date": "Fri, 20 May 2016 02:49:24 GMT",
  "server": "AmazonS3",
  "connection": "close"
}
Make sure your network and proxy settings are correct.
 
If you continue to have issues, please report this full log at https://github.com/Medium/phantomjs
npm ERR! Windows_NT 6.3.9600
npm ERR! argv "C:\\Program Files (x86)\\nodejs\\node.exe" "C:\\Program Files (x86)\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "install" "phantomjs-prebuilt"
npm ERR! node v4.2.6
npm ERR! npm  v2.14.12
npm ERR! code ELIFECYCLE
```

**Comment by les2** [comment]
This works for me. The appveyor build probably needs a re-run to get around a temporary network problem?

```
curl -L https://bitbucket.org/ariya/phantomjs/downloads/phantomjs-2.1.1-windows.zip > phantomjs-2.1.1-windows.zip
```

**Comment by stefanpenner** [comment]
kicking appveyor now

**Comment by les2** [comment]
Pushed a change to a comment to make build run again.

File: lib/tasks/server/middleware/tests-server/index.js (hunk 1 above)
Comment (Turbo87) [inline lib/tasks/server/middleware/tests-server/index.js]: Correct, no need for that code comment though IMHO

File: tests/unit/tasks/server/middleware/tests-server-test.js (hunk 2 above)
Comment (Turbo87) [inline tests/unit/tasks/server/middleware/tests-server-test.js]: I think beforeEach is the right way to go here. Would be awesome if you could change it.

**Comment by Turbo87** [comment]
Looks good in general 👍 

Not sure what's up with CI at the moment...

**Comment by Turbo87** [comment]
@les2 could you rebase on the current master branch? hopefully that should solve the CI issues...

**Comment by les2** [comment]
@Turbo87 rebasing ...

**Comment by les2** [comment]
@Turbo87 @stefanpenner CI passes now

**Comment by Turbo87** [comment]
@les2 awesome, thanks!

@homu r+

**Comment by Turbo87** [comment]
@stefanpenner @nathanhammond I'd like to pull this into the `beta` branch unless there are any objections

**Comment by nathanhammond** [comment]
Late to seeing this. It's in beta now! 😛
